backend/molgenis-emx2/src/main/java/org/molgenis/emx2/utils/generator/SnowFlakeIdGenerator.java

```diff
@@ -0,0 +1,114 @@
+package org.molgenis.emx2.utils.generator;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.math.BigInteger;
+import java.time.Instant;
+import org.molgenis.emx2.MolgenisException;
+
+public class SnowFlakeIdGenerator {
+
+  // It can only have one instance because of colliding ids
+  // So singleton pattern or dependency injection
+  private static SnowFlakeIdGenerator instance;
+
+  private static final String BASE62_CHARACTERS =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+  private static final long EPOCH = 1727446121186L;
+
+  // With base62 encoding 64 bits results in a ~10 character id
+  private static final int TOTAL_BITS = 64;
+  private static final int TIMESTAMP_BITS = 41; // 69 years of ids
+  private static final int SEQUENCE_BITS = 8; // 2^8 (256) ids per ms
+  private static final int ID_BITS =
+      TOTAL_BITS - TIMESTAMP_BITS - SEQUENCE_BITS; // 15 bits left for tableId ~= 32K unique hashes
+
+  private static final long MAX_TIMESTAMP = (1L << TIMESTAMP_BITS) - 1;
+  private static final long MAX_SEQUENCE = (1L << SEQUENCE_BITS) - 1;
+  private static final long MAX_ID = (1L << ID_BITS) - 1;
+
+  // For the schemaId hashing
+  private static final long FNV_OFFSET_BASIS = 0xcbf29ce484222325L;
+  private static final long FNV_PRIME = 0x100000001b3L;
+
+  private long lastTimestamp = -1L;
+  private long sequence = 0L;
+
+  private SnowFlakeIdGenerator() {}
+
+  public static synchronized SnowFlakeIdGenerator getInstance() {
+    if (instance == null) {
+      instance = new SnowFlakeIdGenerator();
+    }
+    return instance;
+  }
+
+  public synchronized String generateId(String tableId) {
+    long currentTimestamp = getCurrentTimestamp();
+
+    if (currentTimestamp == lastTimestamp) {
+      sequence = (sequence + 1) & MAX_SEQUENCE;
+      if (sequence == 0) {
+        currentTimestamp = waitForNextMillis(currentTimestamp);
+      }
+    } else {
+      sequence = 0;
+    }
+    lastTimestamp = currentTimestamp;
+
+    long tableBits = hashStringToBits(tableId);
```

Comment:
I think this was a server id in the original implementation and I think it would be nice to use keep that. Because then we could in the future do federation between multiple EMX2 servers. Since the snowflake generator is a singleton each id generated is already unique over all the schemas in one instance.

Reply:
I think I agree, the table identifier has no added value. Indeed twitter uses worker id (machine/datacenter id). Would you propose to have a unique id per Molgenis instance. Or create a hash from the MAC address maybe?

Reply:
agreed. Table id is not useful, we need a per server uniqueness (or ideallly per server+per schema because conceptually schema is the thing that should be globally unique container)

Reply:
@mswertz yes, but as Esther mentioned, every identifier is inherently unique for each EMX instance. Therefore, adding a table or schema ID to the hash of the Snowflake identifier doesn't provide any additional value. To ensure uniqueness across servers, we could generate a random 16-bit int when the server starts up and store it in the database.